### ISSUE TITLE
Stop using undefined tensors to represent zero gradients in engine

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#4274]
**Motivation.** The treatment of undefined tensors in the autograd engine has been a continual source of confusion in PyTorch internals. Instances of such bugs: #3412, #4215, #4198, #3743. Also see #3801, #2003. (#4217 and #4217 are also undefined tensor bugs but not quite related.) The crux of the problem is that we're using undefined tensors to mean two things:

1. A value for a default initialized tensor, for which we need a non-allocating "null pointer-like" construct suitable for use as the default in optional arguments and all of those places (`std::vector`, `std::tuple`) where C++ makes life very difficult if you don't have a default constructor.

2. When passing around gradients in the autograd engine, we need an efficient mechanism of representing zero gradients.  Semantically, a zero gradient is equivalent to creating a zero filled tensor of the appropriate size and type, but actually materializing such a tensor involves doing a bunch of unnecessary allocations.

These two use-cases are often in tension with each other. For example, it would be no problem to use a null pointer to represent undefined tensors in case (1) (as segfaults in these situations are easy to detect at develop time, and are simply a matter of good hygiene) whereas we have made enough errors in case (2) that we introduced a patch to turn undefined tensors into a full-fledged object, *simply* to avoid segfaulting when we got it wrong; see #3482.

**Proposal.** We propose to *distinguish* between these two use-cases, introducing a new concept of zero-filled tensor. By not eliminating the concept of an undefined tensor, we avoid many of the objections levied against earlier versions of this proposal. There are a number of possible implementation strategies, I'll describe two of them:

* **Zero-filled tensor as immutable zero-dim tensor**
  * For every tensor type, we define a lazily initialized, singleton tensor instance, which is zero-dimensional and zero-filled. As a singleton tensor, there is a pointer-equality test to see if your tensor is zero-filled.
  * This tensor's backing data is placed in read-only memory, so that attempts to "overwrite zero" will result in a fault
  * Autograd engine tests to see if a tensor is zero-filled anywhere it previously tested for undefined tensors, and uses this to skip additions when possible.
  * Can zero filled tensors be passed on to further backwards functions?  The hazard pointed out by @albanD (https://github.com/zdevito/ATen/issues/163#issuecomment-342772547) is that some operations require contiguous tensors. Additionally, it is only valid to pass a zero-dim tensor in place of regular tensor if the operation in question is broadcasting. I see two possible resolutions for this problem:
    * We accept as a contract that we never pass on zero-dimensional gradients to subsequent backward functions. In this case, the engine is expected to materialize the zero-filled tensor into a full tensor if it needs to pass on. This may happen already with InputBuffer.
    * We adjust all backwards functions to broadcast gradients, and all implementations are responsible for expanding and calling contiguous as necessary.
* **Zero-filled tensor as a new Type**
  * The previous strategy is parsimonious, but it has the problem that zero-dim tensors are not semantically equivalent an n-dim zero-filled tensor. An alternate strategy which doesn't have this problem is to introduce another Type which *is* semantically equivalent to an n-dim zero-filled tensor, but with a more efficient backing implementation.
  * As a new Type, we must provide implementations for all of the operations. As a first approximation, all of these operations could be left as exception raising, NYI (similar to how an undefined tensor is defined today), but we could slowly build out support as is necessary.
  * (Not exactly sure how to dispatch CPUDoubleType + ZeroFilledType. @zdevito, maybe you have ideas?)

Tensors continue to default construct into undefined tensors. After this change is in place, we will revert #3482, restoring undefined tensors into actual null pointers which will segfault if you attempt to access them.

This proposal obsoletes https://github.com/zdevito/ATen/issues/163

CC @gchanan @zdevito @colesbury @apaszke @soumith

**Comment by albanD** [comment]
I think that is a very interesting change !

I think I would tend toward the second proposition even though it might be a bit of work.

In particular there is one point that is missing from your comments above (I think):
This point is partially true: `Autograd engine tests to see if a tensor is zero-filled anywhere it previously tested for undefined tensors, and uses this to skip additions when possible.`.
Indeed, in the case where we actually want to create the graph during backward (for higher order gradients), the fact that a gradient is zero does not mean that it's own gradient is going to be 0, so the operation (even if we know the output is going to be 0) must exist in the created graph. By using a specific type, we can easily have the `CPUDoubleTensor * ZeroFilledType` be an autograd op that returns another `ZeroFilledType` and pass it to the rest of the gradient computation.

I got a bit lost on how the dispatch for autograd ops works, but if it is possible to create a `conv` for example that takes `ZeroFilledType` as input as well, we can greatly simplify the implementation of backward functions. There would be no need to check which input is zero or not and some ops will automatically become no-ops when provided with `ZeroFilledType` (while creating the correct backward of backward graph).

Another (minor) advantage would be some slight performance increase for people using `torch.zeros()` to create buffers that they then fill.

**Comment by gchanan** [comment]
> Tensors continue to default construct into undefined tensors. After this change is in place, we will revert #3482, restoring undefined tensors into actual null pointers which will segfault if you attempt to access them.

This seems like an orthogonal concern; we may want to do this because it would allow us to make some functions non-virtual, but that doesn't have to do with how we represent zero-filled tensors in the engine.

**Comment by gchanan** [comment]
going to close this because I believe the issues with undefined tensors in the engine have been addressed.
